File: .tmp_py4013/batch_11_dump/3196_maximize_total_cost_of_alternating_subarrays.js
// LeetCode 3196 - Maximize Total Cost of Alternating Subarrays
// https://leetcode.com/problems/maximize-total-cost-of-alternating-subarrays/

var maximumTotalCost = function(nums) {
    const NEG = -1e18;
    const n = nums.length;
    const memo = Array.from({length: n}, () => [NEG, NEG]);
    const dfs = (i, j) => {
        if (i >= n) return 0;
        if (memo[i][j] !== NEG) return memo[i][j];
        let res = nums[i] + dfs(i + 1, 1);
        if (j > 0) res = Math.max(res, -nums[i] + dfs(i + 1, 0));
        return (memo[i][j] = res);
    };
    return dfs(0, 0);
};

File: .tmp_rs4013/rust_check15/3196_maximize_total_cost_of_alternating_subarrays.rs
struct Solution;
// LeetCode 3196 - Maximize Total Cost of Alternating Subarrays
// https://leetcode.com/problems/maximize-total-cost-of-alternating-subarrays/

impl Solution {
    pub fn maximum_total_cost(nums: Vec<i32>) -> i64 {
        let n = nums.len();
        const NEG: i64 = -1_000_000_000_000_000_000;
        let mut memo = vec![[NEG, NEG]; n];
        fn dfs(i: usize, j: usize, nums: &[i32], memo: &mut [[i64; 2]]) -> i64 {
            if i >= nums.len() {
                return 0;
            }
            if memo[i][j] != -1_000_000_000_000_000_000 {
                return memo[i][j];
            }
            let mut res = nums[i] as i64 + dfs(i + 1, 1, nums, memo);
            if j > 0 {
                res = res.max(-(nums[i] as i64) + dfs(i + 1, 0, nums, memo));
            }
            memo[i][j] = res;
            res
        }
        dfs(0, 0, &nums, &mut memo)
    }
}

fn main() {}
